### PR TITLE
feat(WeightedAuthModule): auth module for the gateway

### PR DESCRIPTION
**Pull request by re1ro**
[AXE-3051]
* [x] `WeightedAuthModule`: new module based on `BaseWeightedMultisig`

[AXE-3051]: https://axelarnetwork.atlassian.net/browse/AXE-3051?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/gateway/WeightedAuthModule.sol [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import { BaseWeightedMultisig } from '../governance/BaseWeightedMultisig.sol';
+import { Ownable } from '../utils/Ownable.sol';
+
+/**
+ * @title WeightedAuthModule Contract
+ * @dev This contract is part of the Axelar Governance system, it inherits the BaseWeightedMultisig contract
+ * with added functionality to approve and execute multisig proposals.
+ */
+contract WeightedAuthModule is Ownable, BaseWeightedMultisig, IWeightedAuthModule {
```
Comment (milapsheth) [inline contracts/gateway/WeightedAuthModule.sol:15]:
```suggestion
contract AxelarVerifiedAuthModule is Ownable, BaseWeightedMultisig, IWeightedAuthModule {
```
perhaps? Other suggestions, `AxelarGatewayWeightedAuth`, `AxelarVerifierAuth`

File: contracts/gateway/WeightedAuthModule.sol [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    // @notice The number of previous signers to authorize batches of commands
+    uint256 public constant PREVIOUS_SIGNERS_RETENTION = 15;
+
+    /**
+     * @notice Initializes the contract.
+     * @dev Ownership of this contract should be transferred to the Gateway contract after deployment.
+     * @param recentOperatorSets The recent operator sets to be added to the multisig as initial signers
+     */
+    constructor(WeightedSigners[] memory recentOperatorSets) Ownable(msg.sender) BaseWeightedMultisig(PREVIOUS_SIGNERS_RETENTION) {
```
Comment (milapsheth) [inline contracts/gateway/WeightedAuthModule.sol:24]:
```suggestion
    constructor(WeightedSigners[] memory initialOperatorSets) Ownable(msg.sender) BaseWeightedMultisig(PREVIOUS_SIGNERS_RETENTION) {
```

File: contracts/gateway/WeightedAuthModule.sol [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import { BaseWeightedMultisig } from '../governance/BaseWeightedMultisig.sol';
+import { Ownable } from '../utils/Ownable.sol';
+
+/**
+ * @title WeightedAuthModule Contract
+ * @dev This contract is part of the Axelar Governance system, it inherits the BaseWeightedMultisig contract
+ * with added functionality to approve and execute multisig proposals.
+ */
+contract WeightedAuthModule is Ownable, BaseWeightedMultisig, IWeightedAuthModule {
+    // @notice The number of previous signers to authorize batches of commands
```
Comment (milapsheth) [inline contracts/gateway/WeightedAuthModule.sol:16]:
```suggestion
    // @notice The number of previous signers whose messages will be considered valid. This gives some time for signed messages to be relayed.
```

File: contracts/gateway/WeightedAuthModule.sol [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+pragma solidity ^0.8.0;
+
+import { IWeightedAuthModule } from '../interfaces/IWeightedAuthModule.sol';
+
+import { BaseWeightedMultisig } from '../governance/BaseWeightedMultisig.sol';
+import { Ownable } from '../utils/Ownable.sol';
+
+/**
+ * @title WeightedAuthModule Contract
+ * @dev This contract is part of the Axelar Governance system, it inherits the BaseWeightedMultisig contract
```
Comment (milapsheth) [inline contracts/gateway/WeightedAuthModule.sol:12]:
```suggestion
 * @dev This contract is used by the Axelar Gateway to verify signed commands. It inherits the BaseWeightedMultisig contract
```

File: contracts/gateway/WeightedAuthModule.sol (hunk 2 above)
Comment (milapsheth) [inline contracts/gateway/WeightedAuthModule.sol:24]: pass in owner as a constructor arg to allow using create2/create3 deployer

Comment (milapsheth) [inline contracts/gateway/WeightedAuthModule.sol:24]: let's use `bytes` for signers similar to validateProof to keep the scripts simpler

File: contracts/gateway/WeightedAuthModule.sol (hunk 1 above)
Comment (re1ro) [reply contracts/gateway/WeightedAuthModule.sol:15]: `AxelarGatewayWeightedAuth`

File: test/gateway/AxelarGatewayWeightedAuth.js [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+                await getWeightedSignaturesProof(
+                    data,
+                    signers,
+                    signers.map(() => 1),
+                    threshold,
+                    signers.slice(0, threshold),
+                ),
+            );
+
+            expect(isCurrentSigners).to.be.equal(true);
```
Comment (deanamiel) [inline test/gateway/AxelarGatewayWeightedAuth.js:64]:
```suggestion
            expect(isCurrentSigners).to.be.true;
```

File: test/gateway/AxelarGatewayWeightedAuth.js [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+                await getWeightedSignaturesProof(
+                    data,
+                    signers,
+                    signers.map(() => 1),
+                    1,
+                    signers.slice(0, 1),
+                ),
+            );
+
+            await expect(isCurrentSigners).to.be.equal(true);
```
Comment (deanamiel) [inline test/gateway/AxelarGatewayWeightedAuth.js:164]:
```suggestion
            await expect(isCurrentSigners).to.be.true;
```

File: test/gateway/AxelarGatewayWeightedAuth.js [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+                        message,
+                        await getWeightedSignaturesProof(
+                            data,
+                            signers,
+                            signers.map(() => index + 2),
+                            (index + 2) * 2,
+                            signers,
+                        ),
+                    );
+                    expect(isCurrentSigners).to.be.equal(false);
```
Comment (deanamiel) [inline test/gateway/AxelarGatewayWeightedAuth.js:213]:
```suggestion
                    expect(isCurrentSigners).to.be.false;
```

File: test/gateway/AxelarGatewayWeightedAuth.js [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+                        ),
+                        gasOptions,
+                    ),
+                multisig,
+                'InvalidSigners',
+            );
+        });
+
+        it('should not allow transferring signership to address zero', async () => {
+            const newSigners = [ethers.constants.AddressZero, '0x6D4017D4b1DCd36e6EA88b7900e8eC64A1D1315b'];
```
Comment (deanamiel) [inline test/gateway/AxelarGatewayWeightedAuth.js:283]:
```suggestion
            const newSigners = [AddressZero, '0x6D4017D4b1DCd36e6EA88b7900e8eC64A1D1315b'];
```

File: test/gateway/AxelarGatewayWeightedAuth.js [hunk 9]
```diff
@@ -0,0 +1,444 @@
+const { sortBy } = require('lodash');
+const chai = require('chai');
+const { ethers, network } = require('hardhat');
+const {
+    utils: { arrayify, keccak256, hashMessage },
```
Comment (deanamiel) [inline test/gateway/AxelarGatewayWeightedAuth.js:5]: Can add `AddressZero` under `constants: { AddressZero }` here

**Review by deanamiel — APPROVED** [review]

**Review by Foivos — APPROVED** [review]

**Review by milapsheth — APPROVED** [review]

**Pull request merged**